strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/repository/group/BaseMavenGroupRepositoryComponent.java

```diff
@@ -0,0 +1,128 @@
+
+    @Inject
+    protected ConfigurationManager configurationManager;
+
+    @Inject
+    protected LayoutProviderRegistry layoutProviderRegistry;
+
+    @Inject
+    protected MavenRepositoryFeatures repositoryFeatures;
+
+    public void updateGroupRepositoriesContainingPathWhenArtifactWasDeleted(final String storageId,
+                                                                            final String repositoryId,
+                                                                            final String artifactRelativePath)
+            throws IOException
+    {
+        final Map<String, MutableBoolean> repositoryArtifactExistence = new HashMap<>();
+        updateGroupRepositoriesContainingPathWhenArtifactWasDeleted(storageId, repositoryId, artifactRelativePath,
+                                                                    repositoryArtifactExistence);
+    }
+
+    private void updateGroupRepositoriesContainingPathWhenArtifactWasDeleted(final String storageId,
+                                                                             final String repositoryId,
+                                                                             final String artifactRelativePath,
+                                                                             final Map<String, MutableBoolean> repositoryArtifactExistence)
+            throws IOException
+    {
+        final List<Repository> directParents = configurationManagementService.getGroupRepositoriesContaining(
+                repositoryId);
+        if (CollectionUtils.isEmpty(directParents))
+        {
+            return;
+        }
+        for (final Repository groupRepository : directParents)
+        {
+            final MutableBoolean artifactExistence = new MutableBoolean();
+            checkArtifactExistenceInTheGroupRepositorySubtree(groupRepository, artifactRelativePath,
+                                                              repositoryArtifactExistence, artifactExistence);
+            if (artifactExistence.isFalse())
+            {
+                performUpdateOfGroupRepositoryContainingPathWhenArtifactWasDeleted(groupRepository,
+                                                                                   artifactRelativePath);
+            }
+            // go higher in the hierarchy
+            updateGroupRepositoriesContainingPathWhenArtifactWasDeleted(groupRepository.getStorage().getId(),
+                                                                        groupRepository.getId(),
+                                                                        artifactRelativePath,
+                                                                        repositoryArtifactExistence);
+        }
+    }
+
+    /**
+     * This method goes down the group repository tree and lookups for an existence of the artifact in the given path.
+     */
+    private void checkArtifactExistenceInTheGroupRepositorySubtree(final Repository groupRepository,
```

Comment:
maybe this can be implemented generic way? I mean that the logic is pretty generic but class called `BaseMavenGroupRepositoryComponent`

Reply:
Thanks. Definietely it requires some code reorganization. It's still in progress. Thanks again.